File: code/turbo_adjudicator.ts
interface Holdings { [address: string]: number; }
interface Allocations { [address: string]: Allocation; }
type Allocation = { address: string; amount: number; }[];

class TurboAdjudicator {
  holdings: Holdings;
  allocations: Allocations;

  constructor() {
    this.holdings = {};
    this.allocations = {};
  }

  register(address: string, allocation: Allocation): boolean {
    if (this.allocations[address]) { return false; }

    this.allocations[address] = allocation;
  }

  transfer(fromAddress: string, toAddress: string, amount: number): boolean {
    if (!this.holdings[fromAddress]) { return false; }
    let fundsAvailable = this.holdings[fromAddress];

    if (!this.allocations[fromAddress]) { return false; }
    const allocation = this.allocations[fromAddress];

    allocation.forEach((item) => {
      if (fundsAvailable < amount) { return false; }
      if (item.address == toAddress) {
        if (item.amount >= amount) {
          this.holdings[fromAddress] -= amount;
          this.holdings[toAddress] += amount;
          item.amount -= amount;
        } else {
          return false;
        }
      }
      fundsAvailable -= item.amount;
    });

    return false;
  }
}

File: code/turbo_adjudicator.rb
class TurboAdjudicator
  def initialize
    @holdings = Hash.new(0)
    @allocations = Hash.new(nil)
  end

  def register(allocation, channelAddress)
    @allocations[channelAddress] ||= allocation
  end

  def transfer(fromAddress, toAddress, amount)
    if canAfford?(fromAddress, toAddress, amount)
      @holdings[fromAddress] += amount
      @holdings[toAddress] -= amount
      @allocations[fromAddress].each do |entry|
        if addr = toAddress
      end
    end
  end

  def canAfford?(fromAddress, toAddress, amount)
    fundsAvailable = @holdings[fromAddress]
    return false if fundsAvailable < amount

    allocation = @allocations[fromAddress]
    return false if allocation.nil?

    allocation.each do |address, total|
      return true if address == toAddress && amount <= total

      fundsAvailable -= total
    end

    return false
  end
end
